File: sdk/node/src/transport.js
// transport.js — the request/reply seam (subprocess pipe or HTTP/1.1).
//
// A transport is `async call(request) -> reply`. Two concrete transports mirror
// the C++/Python SDKs:
//
//   * StdioTransport — newline-delimited JSON over a child's stdin/stdout.
//   * HttpTransport  — minimal HTTP/1.1 POST to `<base>/<method>`.
//
// Failures throw RcpError with the same typed codes as the other SDKs.

import { spawn } from "node:child_process";
import http from "node:http";
import { Errc, RcpError } from "./types.js";

// Newline-delimited JSON over a subprocess' stdio. Replies are correlated to
// requests by JSON-RPC id; server-initiated notifications (frames carrying
// `method`) and stray mismatched-id replies are skipped (spec §4.7).
export class StdioTransport {
  constructor(child) {
    this._child = child;
    this._buf = "";
    this._pending = new Map(); // id -> { resolve, reject }
    this._closed = false;

    child.stdout.setEncoding("utf8");
    child.stdout.on("data", (chunk) => this._onData(chunk));
    child.stdout.on("close", () => this._fail("server closed the connection"));
    child.on("exit", () => this._fail("server exited"));
    child.on("error", (e) => this._fail(`server process error: ${e.message}`));
    child.stdin.on("error", () => {}); // EPIPE surfaces via the write callback
  }

  static spawn(argv) {
    if (!Array.isArray(argv) || argv.length === 0) {
      throw new RcpError(Errc.INVALID_PARAMS, "server command must be a non-empty array");
    }
    let child;
    try {
      child = spawn(argv[0], argv.slice(1), { stdio: ["pipe", "pipe", "inherit"] });
    } catch (e) {
      throw new RcpError(Errc.BACKEND_UNAVAILABLE, `cannot spawn ${argv[0]}: ${e.message}`);
    }
    return new StdioTransport(child);
  }

  call(request) {
    return new Promise((resolve, reject) => {
      if (this._closed) {
        reject(new RcpError(Errc.BACKEND_UNAVAILABLE, "transport is closed"));
        return;
      }
      const id = request.id;
      this._pending.set(id, { resolve, reject });
      this._child.stdin.write(JSON.stringify(request) + "\n", (err) => {
        if (err) {
          this._pending.delete(id);
          reject(new RcpError(Errc.BACKEND_UNAVAILABLE, "write to server failed"));
        }
      });
    });
  }

  _onData(chunk) {
    this._buf += chunk;
    let nl;
    while ((nl = this._buf.indexOf("\n")) >= 0) {
      const line = this._buf.slice(0, nl);
      this._buf = this._buf.slice(nl + 1);
      if (!line.trim()) continue;
      let reply;
      try {
        reply = JSON.parse(line);
      } catch {
        continue; // skip a line we can't parse rather than derail the stream
      }
      if (reply && typeof reply === "object" && "method" in reply) continue; // notification
      const waiter = this._pending.get(reply?.id);
      if (waiter) {
        this._pending.delete(reply.id);
        waiter.resolve(reply);
      }
      // else: a stray reply with an unknown id — ignore.
    }
  }

  _fail(message) {
    if (this._closed) return;
    this._closed = true;
    const err = new RcpError(Errc.BACKEND_UNAVAILABLE, message);
    for (const { reject } of this._pending.values()) reject(err);
    this._pending.clear();
  }

  close() {
    this._closed = true;
    try { this._child.stdin.end(); } catch {}
    try { this._child.kill("SIGTERM"); } catch {}
  }
}

// Minimal HTTP/1.1 client: POST the request to `<baseUrl>/<method>`. The server
// ignores the path and dispatches on the body's `method`; the path is a
// human-readable convenience. Maps HTTP 429 -> RateLimited and 503 ->
// BackendUnavailable (spec §5.2).
export class HttpTransport {
  constructor(baseUrl, timeoutMs = 30000) {
    this._base = String(baseUrl).replace(/\/+$/, "");
    this._timeout = timeoutMs;
  }

  call(request) {
    return new Promise((resolve, reject) => {
      const method = request.method || "";
      let url;
      try {
        url = new URL(this._base + "/" + method);
      } catch (e) {
        reject(new RcpError(Errc.INVALID_PARAMS, `bad URL: ${e.message}`));
        return;
      }
      if (url.protocol !== "http:") {
        reject(new RcpError(Errc.INVALID_PARAMS, `unsupported URL scheme '${url.protocol}'`));
        return;
      }
      const body = Buffer.from(JSON.stringify(request), "utf8");
      const req = http.request(
        {
          hostname: url.hostname,
          port: url.port || 80,
          path: url.pathname + url.search,
          method: "POST",
          headers: { "Content-Type": "application/json", "Content-Length": body.length },
          timeout: this._timeout,
        },
        (res) => {
          const chunks = [];
          res.on("data", (c) => chunks.push(c));
          res.on("end", () => {
            const status = res.statusCode;
            const data = Buffer.concat(chunks);
            if (status === 429) { reject(new RcpError(Errc.RATE_LIMITED, "HTTP 429")); return; }
            if (status === 503) { reject(new RcpError(Errc.BACKEND_UNAVAILABLE, "HTTP 503")); return; }
            if ((status < 200 || status >= 300) && data.toString("utf8").trim() === "") {
              reject(new RcpError(Errc.BACKEND_UNAVAILABLE, `HTTP ${status}`));
              return;
            }
            try {
              resolve(JSON.parse(data.toString("utf8")));
            } catch (e) {
              reject(new RcpError(Errc.PARSE_ERROR, e.message));
            }
          });
        }
      );
      req.on("error", (e) => reject(new RcpError(Errc.BACKEND_UNAVAILABLE, `HTTP request failed: ${e.message}`)));
      req.on("timeout", () => req.destroy(new RcpError(Errc.BACKEND_UNAVAILABLE, "HTTP request timed out")));
      req.write(body);
      req.end();
    });
  }

  close() {}
}

File: sdk/rust/src/transport.rs
//! transport.rs — the request/reply seam (subprocess pipe or HTTP/1.1).
//!
//! A `Transport` is a total function `&Json -> Result<Json, RcpError>`. Two
//! concrete transports mirror the other SDKs, both built on `std` alone:
//!
//!   * `StdioTransport` — newline-delimited JSON over a child's stdin/stdout.
//!   * `HttpTransport`   — minimal HTTP/1.1 POST to `<base>/<method>`.

use crate::json::Json;
use crate::types::{Errc, RcpError};
use std::io::{BufRead, BufReader, Read, Write};
use std::net::TcpStream;
use std::process::{Child, ChildStdin, ChildStdout, Command, Stdio};

/// One JSON-RPC request → one reply.
pub trait Transport {
    fn call(&mut self, request: &Json) -> Result<Json, RcpError>;
    fn close(&mut self);
}

/// Newline-delimited JSON over a subprocess' stdio. Replies are correlated by
/// JSON-RPC id; server notifications (frames carrying `method`) and stray
/// mismatched-id replies are skipped (spec §4.7).
pub struct StdioTransport {
    child: Child,
    stdin: ChildStdin,
    stdout: BufReader<ChildStdout>,
}

impl StdioTransport {
    pub fn spawn(argv: &[String]) -> Result<StdioTransport, RcpError> {
        if argv.is_empty() {
            return Err(RcpError::new(
                Errc::INVALID_PARAMS,
                "server command must be non-empty",
            ));
        }
        let mut child = Command::new(&argv[0])
            .args(&argv[1..])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            // stderr inherits: server diagnostics reach the launching terminal.
            .spawn()
            .map_err(|e| {
                RcpError::new(
                    Errc::BACKEND_UNAVAILABLE,
                    format!("cannot spawn {}: {}", argv[0], e),
                )
            })?;
        let stdin = child.stdin.take().expect("piped stdin");
        let stdout = BufReader::new(child.stdout.take().expect("piped stdout"));
        Ok(StdioTransport {
            child,
            stdin,
            stdout,
        })
    }
}

impl Transport for StdioTransport {
    fn call(&mut self, request: &Json) -> Result<Json, RcpError> {
        let mut line = request.dump();
        line.push('\n');
        self.stdin
            .write_all(line.as_bytes())
            .and_then(|_| self.stdin.flush())
            .map_err(|_| RcpError::new(Errc::BACKEND_UNAVAILABLE, "write to server failed"))?;

        let want_id = request.get("id").cloned();
        loop {
            let mut buf = String::new();
            let n = self.stdout.read_line(&mut buf).map_err(|e| {
                RcpError::new(Errc::BACKEND_UNAVAILABLE, format!("read failed: {}", e))
            })?;
            if n == 0 {
                return Err(RcpError::new(
                    Errc::BACKEND_UNAVAILABLE,
                    "server closed the connection",
                ));
            }
            let trimmed = buf.trim();
            if trimmed.is_empty() {
                continue;
            }
            let reply = Json::parse(trimmed).map_err(|e| RcpError::new(Errc::PARSE_ERROR, e))?;
            if reply.contains_key("method") {
                continue; // server-initiated notification
            }
            if let (Some(want), Some(got)) = (&want_id, reply.get("id")) {
                if want != got {
                    continue; // stray reply for an earlier request
                }
            }
            return Ok(reply);
        }
    }

    fn close(&mut self) {
        let _ = self.stdin.flush();
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

impl Drop for StdioTransport {
    fn drop(&mut self) {
        // Never orphan the child: std's Child::drop detaches by default.
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

/// Minimal HTTP/1.1 client: POST the request to `<base_url>/<method>`. The server
/// ignores the path and dispatches on the body's `method`. Maps HTTP 429 →
/// RateLimited and 503 → BackendUnavailable (spec §5.2).
pub struct HttpTransport {
    base: String,
}

impl HttpTransport {
    pub fn new(base_url: &str) -> HttpTransport {
        HttpTransport {
            base: base_url.trim_end_matches('/').to_string(),
        }
    }
}

impl Transport for HttpTransport {
    fn call(&mut self, request: &Json) -> Result<Json, RcpError> {
        let method = request.get_str("method").unwrap_or("");
        let url = format!("{}/{}", self.base, method);
        let (host, port, path) = parse_http_url(&url)?;
        let body = request.dump();
        let req = format!(
            "POST {} HTTP/1.1\r\nHost: {}\r\nContent-Type: application/json\r\n\
             Content-Length: {}\r\nConnection: close\r\n\r\n{}",
            path,
            host,
            body.len(),
            body
        );
        let mut stream = TcpStream::connect((host.as_str(), port))
            .map_err(|e| RcpError::new(Errc::BACKEND_UNAVAILABLE, format!("connect failed: {}", e)))?;
        stream
            .write_all(req.as_bytes())
            .map_err(|e| RcpError::new(Errc::BACKEND_UNAVAILABLE, format!("write failed: {}", e)))?;
        let mut resp = Vec::new();
        stream
            .read_to_end(&mut resp)
            .map_err(|e| RcpError::new(Errc::BACKEND_UNAVAILABLE, format!("read failed: {}", e)))?;

        let (status, body_bytes) = parse_http_response(&resp)?;
        if status == 429 {
            return Err(RcpError::new(Errc::RATE_LIMITED, "HTTP 429"));
        }
        if status == 503 {
            return Err(RcpError::new(Errc::BACKEND_UNAVAILABLE, "HTTP 503"));
        }
        let text = String::from_utf8_lossy(body_bytes);
        if !(200..300).contains(&status) && text.trim().is_empty() {
            return Err(RcpError::new(
                Errc::BACKEND_UNAVAILABLE,
                format!("HTTP {}", status),
            ));
        }
        Json::parse(text.trim()).map_err(|e| RcpError::new(Errc::PARSE_ERROR, e))
    }

    fn close(&mut self) {}
}

/// Split `http://host[:port]/path` into `(host, port, path)`.
fn parse_http_url(url: &str) -> Result<(String, u16, String), RcpError> {
    let rest = url
        .strip_prefix("http://")
        .ok_or_else(|| RcpError::new(Errc::INVALID_PARAMS, "only http:// URLs are supported"))?;
    let (authority, path) = match rest.find('/') {
        Some(i) => (&rest[..i], &rest[i..]),
        None => (rest, "/"),
    };
    let (host, port) = match authority.rsplit_once(':') {
        Some((h, p)) => (
            h.to_string(),
            p.parse::<u16>()
                .map_err(|_| RcpError::new(Errc::INVALID_PARAMS, "bad port"))?,
        ),
        None => (authority.to_string(), 80),
    };
    Ok((host, port, path.to_string()))
}

/// Return `(status_code, body_slice)` from a raw HTTP/1.1 response.
fn parse_http_response(resp: &[u8]) -> Result<(u16, &[u8]), RcpError> {
    let sep = find_subslice(resp, b"\r\n\r\n")
        .ok_or_else(|| RcpError::new(Errc::BACKEND_UNAVAILABLE, "malformed HTTP response"))?;
    let head = &resp[..sep];
    let body = &resp[sep + 4..];
    // Status line: "HTTP/1.1 200 OK"
    let line_end = find_subslice(head, b"\r\n").unwrap_or(head.len());
    let status_line = String::from_utf8_lossy(&head[..line_end]);
    let status = status_line
        .split_whitespace()
        .nth(1)
        .and_then(|s| s.parse::<u16>().ok())
        .ok_or_else(|| RcpError::new(Errc::BACKEND_UNAVAILABLE, "no HTTP status"))?;
    Ok((status, body))
}

fn find_subslice(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    haystack
        .windows(needle.len())
        .position(|w| w == needle)
}
